src/postgres-util/src/desc.rs

```diff
@@ -28,6 +33,110 @@ pub struct PostgresTableDesc {
     pub name: String,
     /// The description of each column, in order of their position in the table.
     pub columns: Vec<PostgresColumnDesc>,
+    /// Applicable keys for this table (i.e. primary key and unique
+    /// constraints).
+    pub keys: BTreeSet<PostgresKeyDesc>,
+}
+
+impl PostgresTableDesc {
+    /// Determines if two `PostgresTableDesc` are compatible with one another in
+    /// a way that Materialize can handle.
+    ///
+    /// Currently this means that the values are equal except for the following
+    /// exceptions:
+    /// - `self`'s columns are a compatible prefix of `other`'s columns.
+    ///   Compatibility is defined as returning `true` for
+    ///   `PostgresColumnDesc::is_compatible`.
+    /// - `self`'s keys are all present in `other`
+    pub fn determine_compatibility(&self, other: &PostgresTableDesc) -> Result<(), anyhow::Error> {
+        if self == other {
+            return Ok(());
+        }
+
+        let PostgresTableDesc {
+            oid: other_oid,
+            namespace: other_namespace,
+            name: other_name,
+            columns: other_cols,
+            keys: other_keys,
+        } = other;
+
+        // Table columns cannot change position, so only need to ensure that
+        // `self.columns` is a prefix of `other_cols`.
+        if self.columns.len() == other_cols.len()
+            && self.columns.iter().zip(other_cols.iter()).all(|(s, o)| s.is_compatible(o))
+            && &self.name == other_name
+            && &self.oid == other_oid
+            && &self.namespace == other_namespace
+            // Our keys are all still present in exactly the same shape.
+            && self.keys.difference(other_keys).next().is_none()
+        {
+            Ok(())
+        } else {
+            warn!(
+                "Error validating table in publication. Expected: {:?} Actual: {:?}",
+                &self, other
+            );
+            bail!(
+                "source table {} with oid {} has been altered",
+                self.name,
+                self.oid
+            )
+        }
+    }
+
+    /// Determines if `self` is compatible with a [`RelationBody`] in a way that
+    /// Materialize can handle. For details about the return value, see
+    /// [`Self::determine_compatibility`] whose semantics should remain
+    /// identical.
+    ///
+    /// Note that we perform this check locally just to fail eagerly; we must
+    /// reach out to the upstream PG source to get more information about the
+    /// table's current state even if this succeeds.
+    pub fn determine_compatibility_w_relation_body(
+        &self,
+        relation: &RelationBody,
```

Comment:
Noting this from our call. We should consider only doing the check after reaching out to postgres if that's less code to maintain and think about since the expectation is that Relation messages will be rare

Reply:
Removed this check.